Smithy Protocol Test Generator

peterrsongg opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
This PR includes the Java plugin for the protocol test generator and 1 of each type of protocol test (request/response for each protocol, and errrorResponse). The protocol tests are defined in the [smithy repository] (https://github.com/smithy-lang/smithy/tree/main/smithy-aws-protocol-tests/model). The protocol tests are generated from smithy, but the structures and shapes used in the protocol tests are generated from C2J. 

For the member of the smithy team reviewing this, rather than reviewing the generated C# code, I think a review of the java plugin for best practices / possible inefficiencies would be the best use of your time. We were guided not to use DirectedCodegen, since we are just building a Protocol Test Generator. If I'm using an api incorrectly or there is a more efficient way to do things, please comment on the PR.

For the .NET team, a look at the generated code / custom code in the protocol tests project is sufficient, but if you would like to review the java plugin that would be appreciated as well, but not required.

**Please ignore all the gradlew stuff and the smithy-dotnet-codegen-test project**

**NOTE:** Running and iterating on the broken protocols is not part of this task. We are merging into a feature branch if approved.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open [issue][issues], please link to the issue here -->
The .NET SDK has been around before the smithy protocol tests were defined, so while other smithy-based SDKs have onboarded to the smithy protocol tests, we have not. There are numerous cases defined in the internal smithy protocol tests design document which could've been avoided if we had implemented protocol tests. Due to it being an internal document I cannot paste it here.

## Testing
<!--- Please describe in detail how you tested your changes -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Ran the protocol test generator to view the generated C# code and fixed any incorrectly generated test cases by tweaking the plugin. At this point I feel the test generator is close to complete

## Screenshots (if appropriate)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have read the **README** document
- [x] I have added tests to cover my changes
- [ ] All new and existing tests passed

## License
<!--- The SDK is released under the [Apache 2.0 license][license], so any code you submit will be released under that license -->
<!--- For substantial contributions, we may ask you to sign a [Contributor License Agreement (CLA)][cla] -->
<!--- Put an `x` in the below box if you confirm that this request can be released under the Apache 2 license -->
- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

ashovlin commented on sdk/test/ProtocolTests/Generated/QueryProtocol/QueryLists.cs:
```diff
@@ -0,0 +1,232 @@
+        [TestCategory("ProtocolTest")]
+        [TestCategory("RequestTest")]
+        [TestCategory("AwsQuery")]
+        public void FlattenedQueryListsRequest()
```
Assuming that this corresponds to https://github.com/smithy-lang/smithy/blob/0284c86dcb910034d3be84c3fcd6b201790cbe81/smithy-aws-protocol-tests/model/awsQuery/input-lists.smithy#L60, can you use the modeled `documentation` as an XML comment to make it clear to a reader what the test is exercising?

ashovlin commented on sdk/test/ProtocolTests/Generated/QueryProtocol/QueryLists.cs:
```diff
@@ -0,0 +1,232 @@
+                    },
+                },
+            };
+            var config = new AmazonQueryProtocolConfig{
+              ServiceURL = "https://test.com/"
+            };
```
Style throughout isn't consistent with other generated C# code in the repo, e.g.
* Opening braces are usually on next line, not same line
* Missing new lines between separate methods

ashovlin commented:
Can you add a new README in `generator/ProtocolTestsGenerator` covering how to install prerequisites and how to run the generator?

peterrsongg commented:
@ashovlin Thanks for the suggestions. I went ahead and added a READM as well as the test documentation as XML comments. I also broke out each test into Arrange vs Act vs Assert and fixed other styling details

normj commented on generator/ProtocolTestsGenerator/README.md:
````diff
@@ -0,0 +1,39 @@
+```
+2. Build the project
+```
+./gradlew :smithy-dotnet-protocol-test:build
````
The `gradew.bat` is in the parent directory of `smithy-dotnet-protocol-test`. I think this should be 
```
../gradlew :smithy-dotnet-protocol-test:build
```

peterrsongg replied:
Ah, that is a mistake on my end. You're right, you should be in the parent directory. Updated the README to reflect that.

normj commented on generator/ProtocolTestsGenerator/README.md:
````diff
@@ -0,0 +1,39 @@
+```
+2. Build the project
+```
+./gradlew :smithy-dotnet-codegen-test:build
````
I'm getting the following error on this step. Keep in mind per my previous comment I'm using `../gradlew`.

```
Projection source failed: software.amazon.smithy.model.node.ExpectationNotMetException: Missing expected member `package`. (C:\codebase\v3\AWSDotNetPublic\generator\ProtocolTestsGenerator\smithy-dotnet-codegen-test\smithy-build.json [4, 27])
software.amazon.smithy.model.node.ExpectationNotMetException: Missing expected member `package`. (C:\codebase\v3\AWSDotNetPublic\generator\ProtocolTestsGenerator\smithy-dotnet-codegen-test\smithy-build.json [4, 27])
```

peterrsongg replied:
This error is because I updated the `DotnetSettings.java` file to expect different values in the `smithy-build.json` and I forgot to update the test files. I updated the smithy-build.json for the test project. However, running the test project will throw an exception saying "The given protocol is not supported" which is because the core smithy code generator is configured to specifically generate protocol tests. I added a note in the README which warns of this. Since we're not a full smithy code generator, the test project needs some tweaking to work normally.

gosar commented on sdk/test/ProtocolTests/Generated/EC2Protocol/GreetingWithErrors.cs:
Is stuff under `sdk/test/ProtocolTests/Generated` where you are checking in the generated output from the Smithy plugin? As a way to bring in the generated code in your IDE to see/run? i.e., this is temporary and copying generated tests into repo is not expected down the line?

peterrsongg replied:
I didn't know of a way to tell smithy to output the files somewhere, so currently I just have a script that copies the files from the `build` directory to the `sdk/test/ProtocolTests/Generated` directory. But to answer your question, we do expect to copy the generated tests into the repo down the line since we want to create a `csproj` containing all the tests and run them during each release. This also allows devs to easily check protocol behavior without having to run the generator themselves.

peterrsongg replied:
I just found how to override the output directory while looking at another link you sent in a comment, I'll update this too

gosar reviewed (commented):
Publishing one round of comments with what I went through so far - mostly related to setup. Looking at the plugin Java code next.

peterrsongg commented on generator/ProtocolTestsGenerator/build.gradle.kts:
```diff
@@ -0,0 +1,20 @@
+
+repositories {
+    mavenCentral()
+    mavenLocal()
```
thank you I'll change that

peterrsongg commented on generator/ProtocolTestsGenerator/smithy-dotnet-protocol-test/smithy-build.json:
```diff
@@ -0,0 +1,160 @@
+        }
+      },
+        {
+          "name": "excludeShapesBySelector",
```
The smithy team provided some guidance on tests that we should leave out either due to issues with smithy2c2j or because the smithy2c2j artifact isn't up to date with the latest smithy release.

gosar replied:
I'm wondering about adding that context as comment here, so someone reviewing it later can know and also evaluate if it is time to reduce this list. Managing that in the smithy-build.json file probably isn't great. Could this be managed in ProtocolTestCustomization where you have logic to skip other things, and then you can have comment/context/TODOs there?

peterrsongg replied:
I wanted to separate out the tests where entire operations are skipped and cases where we just skip invididual test id's. I thought if smithy provided these exclusion selectors I should use them. But, I can see how it would be hard to maintain why we skipped them later down the line. I'll move them.

peterrsongg commented on generator/ProtocolTestsGenerator/smithy-dotnet-protocol-test/smithy-build.json:
```diff
@@ -0,0 +1,160 @@
+    "RestJsonProtocol": {
+      "transforms": [
+        {
+        "name": "includeServices",
```
fixed

gosar commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/HttpProtocolTestGenerator.java:
```diff
@@ -0,0 +1,510 @@
+    }
+
+    private final class ValueNodeVisitor implements NodeVisitor<Void> {
+        private final Shape inputShape;
```
wherever it says inputShape, maybe say just `shape` as it can be input or output

peterrsongg replied:
I think of this as the shape passed (input) into the ValueNodeVisitor. I think I'll leave this one as is

gosar commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/HttpProtocolTestGenerator.java:
```diff
@@ -0,0 +1,510 @@
+                    getDocument((DocumentShape) inputShape, node);
+                    return null;
+                case UNION:
+                    getUnion((UnionShape) inputShape, node);
```
better to include `return null` here too, to avoid risk or another case getting added and forgetting to return from the UNION case.

gosar commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/HttpProtocolTestGenerator.java:
```diff
@@ -0,0 +1,510 @@
+        writer.write("Assert.AreEqual((HttpStatusCode)Enum.ToObject(typeof(HttpStatusCode), $L), context.ResponseData.StatusCode);",httpResponseTestCase.getCode());
+    }
+
+    private final class ValueNodeVisitor implements NodeVisitor<Void> {
```
i didn't pay super attention to this as it depends on how types are converted in C# which I'm not familiar with

gosar commented on generator/ProtocolTestsGenerator/settings.gradle.kts:
```diff
@@ -0,0 +1,27 @@
+            library("junit-jupiter", "org.junit.jupiter:junit-jupiter:5.9.1")
+        }
+    }
+}
```
nit: newline at end of file. there should be an IDE setting that can help with this, so it happens to all files.

gosar commented on generator/ProtocolTestsGenerator/settings.gradle.kts:
```diff
@@ -0,0 +1,27 @@
+    }
+}
+dependencyResolutionManagement {
+    versionCatalogs {
```
while i've seen it using .toml files in our other repos, this seems fine

gosar commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/HttpProtocolTestGenerator.java:
```diff
@@ -0,0 +1,510 @@
+    }
+
+    private void arrangeResponseTestBlock(HttpResponseTestCase httpResponseTestCase){
+        writer.addImport(protocolNamespace,"System.Net");
+        writer.writeSingleLineComment("Arrange");
+        writer.write("var webResponseData = new WebResponseData();");
```
I'm thinking about how to manage imports. I assume System.Net is imported for WebResponseData, but that is unclear reading this code. So it is difficult for a code writer/reviewer to know that all the needed imports are added. Is there are way to do something like `addImport("System.Net", "WebResponseData")` which will cause the `using System.Net` to be added?

peterrsongg replied:
yup, I will make this change to make it clearer what the imports are for. I had the logic for this in DotnetImportContainer but I never populated it in the CSharpWriter.

peterrsongg commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/DotnetClientCodegenPlugin.java:
```diff
@@ -0,0 +1,84 @@
+    private static HttpProtocolTestGenerator getHttpProtocolTestGenerator(PluginContext context, DotnetGenerationContext generationContext) {
+        var projectionName = context.getProjectionName();
+        HttpProtocolTestGenerator testGenerator;
+        if (projectionName.toLowerCase().contains("json")){
```
That's a good point, I think initially I didn't want to bloat the `DotnetGenerationContext` with this information, but I see no reason not to include the projection name as a property of the `DotnetGenerationContext` class. That way I don't need to couple the logic between the `DotnetClientCodegenPlugin` and the `HttpProtocolTestGenerator`. I will fix this by adding additional info to the generation context and passing only that in to the test generator.

gosar replied:
> I see no reason not to include the projection name as a property of the `DotnetGenerationContext` class

Maybe, but I think using projectionName for any logic doesn't seem great. Can't you achieve it by looking at protocol trait?

peterrsongg replied:
I like that idea better better actually. Will update

peterrsongg commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/DotnetDelegator.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Manages writers for C# Files
+ */
+public class DotnetDelegator extends WriterDelegator<CSharpWriter> {
```
Good callout, will fix

peterrsongg commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/HttpProtocolTestGenerator.java:
```diff
@@ -0,0 +1,510 @@
+            DotnetGenerationContext context, String protocol, String projectionName
+            ) {
+        this.settings = context.settings();
+        this.protocolNamespace = context.settings().getService().getName();
```
The name of the service corresponds to the protocol, but it is a bit confusing. Fixed

peterrsongg commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/DotnetImportContainer.java:
```diff
@@ -0,0 +1,116 @@
+    public void addMarshallImports(String namespace, String service) {
+        addImport(namespace,service);
+        addImport(namespace, service + ".Model");
+        addImport(namespace, service + ".Model.Internal.MarshallTransformations");
```
I decided to put this in here because in the .NET SDK all marshallers/unmarshallers live in the `.Model.Internal.MarshallTransformations` namespace. So I thought this is something we could re-use for a full smithy codegenerator later. i.e. make it generic and not specific to protocol tests.

peterrsongg commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/CSharpWriter.java:
```diff
@@ -0,0 +1,155 @@
+     * <a href = https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions#style-guidelines>C# Style Guidelines</a>
+     */
+    public String formatDocs(String docs){
+        return StringUtils.wrap(docs.replace("\n", " "),65);
```
Yup 65 is the recommended line break by microsoft because of mobile

peterrsongg commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/utils/ProtocolTestUtils.java:
```diff
@@ -0,0 +1,34 @@
+        String namespace = settings.getPackageNamespace();
+        String[] packageNamespaceSplit = namespace.split("\\.");
+        if(packageNamespaceSplit.length != 2){
+            throw new CodegenException("The packageNamespace format in the smithy-build.json file is in the wrong format. Please make sure that is in the format" +
```
I just put this check here for consistency. For example if the length was greater or less than 2, then we'd be using the wrong name for the Marshall Imports. The packageNamespace is used as the base namespace for the test services.

For example, if the packageNamespace is `Amazon.RestJsonProtocol`, we'd correctly import `Amazon.RestJsonProtocol.Model` etc..
but if the length > 2, then we might import something incorrectly like `Amazon.RestJsonProtocol.Foo.Model` which doesn't exist.

gosar replied:
I think this value would be derived from https://github.com/smithy-lang/smithy/blob/2479286123d79c1ab96981a68a6c7c43b0629f82/smithy-aws-protocol-tests/model/awsJson1_0/main.smithy#L11. Eventually when the full client generator is moved to Smithy instead of C2J, you'd need to use this to compute the packageNamespace. So I think you should remove it from Settings and add logic to compute this value from the model.

peterrsongg replied:
Ah you're right that is a cleaner way to handle this, I'll update this

peterrsongg commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/DotnetImportContainer.java:
```diff
@@ -0,0 +1,116 @@
+        // Only import symbols in other namespaces.
+        if (!symbolNamespace.equals(localNamespace)) {
+            var namespaceImports = imports.computeIfAbsent(symbolNamespace, ns -> new TreeMap<>());
+            namespaceImports.put(alias, symbol.getName());
```
It should. fixed.

peterrsongg commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/DotnetImportContainer.java:
```diff
@@ -0,0 +1,116 @@
+            "Amazon.Runtime.Internal.Transform"
+    );
+    private static final List<String> protocolTestImports = Arrays.asList(
+            "AWSSDK.ProtocolTests",
```
Since we updated the plugin to refer to protocol tests, I think this is fine now

peterrsongg commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/Templates.java:
```diff
@@ -0,0 +1,50 @@
+
+package software.amazon.smithy.dotnet.codegen;
+
+public final class Templates {
```
I wanted to mimic the idea of a static class, which exists in C# but not in Java. I found I could do this with a private constructor and a class declared as final. I wanted to separate out these large strings as to not distract from the CSharpWriter, and to leave this here in case there are more things we want to add. But my intention was to just hold constant strings

peterrsongg commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/SymbolVisitor.java:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    @Override
+    public Symbol blobShape(BlobShape blobShape) {
```
Currently this is the behavior in the .NET SDK. If it is marked with @streaming and is a blob type we create a memoryStream to hold the data

peterrsongg commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/DotnetSettings.java:
```diff
@@ -0,0 +1,119 @@
+    private void setPackageNamespace(String packageNamespace) {
+        this.packageNamespace = Objects.requireNonNull(packageNamespace);
+    }
+    private void setService(ShapeId service) {
```
I think I'll keep this as a class, but I moved the getter

peterrsongg commented on sdk/test/ProtocolTests/Generated/JSONRPC10/GreetingWithErrors.cs:
```diff
@@ -0,0 +1,332 @@
+            var stream = new MemoryStream(bytes);
+            var context = new JsonUnmarshallerContext(stream,true,webResponseData);
+            // Act
+            var errorResponse = new GreetingWithErrorsResponseUnmarshaller().UnmarshallException(context, new InvalidGreetingException(""), (HttpStatusCode)Enum.ToObject(typeof(HttpStatusCode), 400));
```
I'm not sure how other SDK's handle errors but we just take the status code that the service returns and either return the specific exception (if it can be determined by the headers and other data in the payload) along with the message, status code, and inner exceptio etc. or if that can't be determined we return the service-level exception. The expected exception is there to return as the inner exception of the parent exception.

For the purposes of the test, I had to pass in this data because the unmarshallException method expects them.

gosar replied:
offline we talked about:
* 2nd parameter being `innerException`. Can avoid using the same exception type as the expected exception test of this test (i.e., InvalidGreetingException) and instead use a generic runtime exception or null if that's allowed.
* 3rd parameter: i'd imagine the unmarshaller is eventually setting errorResponse.statusCode based on WebResponseData in the 1st method parameter OR this explicit value in 3rd parameter. It seems that currently the 3rd parameter is getting used and not WebResponseData. You could see if you'd like to change that in your Unmarshaller logic itself. But maybe fine to set the value in both WebResponseData and this 3rd parameter, so you can change the Unmarshaller later without affecting this test.
  * But I'd say either ways, add an assert for errorResponse.statusCode == 400

gosar replied:
without more context, it seems here https://github.com/aws/aws-sdk-net/blob/23b74540be451eda8bb789980b4863bd8cc4241f/sdk/src/Core/Amazon.Runtime/Internal/Transform/JsonErrorResponseUnmarshaller.cs#L133 this might have worked `response.StatusCode = context.ResponseData.StatusCode` and avoided the 3rd parameter to the method

peterrsongg replied:
I decided to go with the 2nd parameter being null. It seems like the 3rd parameter is being used and not WebResponseData, but I'd like to change the unmarshaller after the fact and not change the test. The test works as expected with the 2nd parameter being null though

gosar commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/HttpProtocolTestGenerator.java:
```diff
@@ -0,0 +1,510 @@
+        writer.writeSingleLineComment("Arrange");
+        writer.write("var webResponseData = new WebResponseData();");
+        writer.write("webResponseData.StatusCode = (HttpStatusCode)Enum.ToObject(typeof(HttpStatusCode), $L);",httpResponseTestCase.getCode());
+        for(var header: httpResponseTestCase.getHeaders().keySet()){
```
required/forbidHeaders check still seems missing from error response tests

gosar commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/HttpProtocolTestGenerator.java:
```diff
@@ -0,0 +1,510 @@
+            writer.write("Assert.AreEqual($S.Replace(\" \",\"\"), marshalledRequest.Headers[$S].Replace(\" \",\"\"));", headers.get(header), header);
+        }
+        // Verify the query Params.
+        if(!httpRequestTestCase.getQueryParams().isEmpty()){
```
requireQueryParams, forbidQueryParams checks still seem missing

gosar commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/HttpProtocolTestGenerator.java:
```diff
@@ -0,0 +1,526 @@
+        writer.write("Assert.AreEqual((HttpStatusCode)Enum.ToObject(typeof(HttpStatusCode), $L), context.ResponseData.StatusCode);", httpResponseTestCase.getCode());
+    }
+
+    private void getMarshallerType(String protocol) {
```
'set'MarshallerType?

gosar commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/HttpProtocolTestGenerator.java:
```diff
@@ -0,0 +1,526 @@
+
+    private void getMarshallerType(String protocol) {
+        if (protocol.toLowerCase().contains("json")) {
+            this.marshallerType = "Json";
```
non-blocking: could make this an internal enum instead of strings

gosar reviewed (commented):
There are still many files with missing newline at end of file

peterrsongg commented on generator/ProtocolTestsGenerator/smithy-dotnet-protocol-test/smithy-build.json:
```diff
@@ -0,0 +1,145 @@
+        }
+      }
+    },
+    "JSONRPC10": {
```
Yup this value is tied to the C2J i.e. the name of the service. I left this in here on purpose to match

gosar replied:
Well, you don't depend on the projectionName value anymore, but on the packageNamespace value from Settings.

gosar replied:
But see this comment https://github.com/aws/aws-sdk-net/pull/3230#discussion_r1561907600

gosar commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/DotnetSettings.java:
```diff
@@ -0,0 +1,124 @@
+    public static DotnetSettings from(ObjectNode config) {
+        DotnetSettings settings = new DotnetSettings();
+        settings.setService(config.expectStringMember(SERVICE).expectShapeId());
+        settings.setPackageName(config.expectStringMember(PACKAGE_NAME).getValue());
+        settings.setPackageVersion(config.expectStringMember(PACKAGE_VERSION).getValue());
+        settings.setPackageNamespace(config.expectStringMember(PACKAGE_NAMESPACE).getValue());
```
IIUC, only packageNamespace is used currently. I suggest removing packageName and packageVersion for now until you have usage for it. And see https://github.com/aws/aws-sdk-net/pull/3230#discussion_r1561907600 for removing packageNamespace.

peterrsongg replied:
Per your suggestion I completely dropped these two and am using the sdkId on the service trait to derive these values.

peterrsongg commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/HttpProtocolTestGenerator.java:
```diff
@@ -0,0 +1,510 @@
+        //Arrange
+        arrangeResponseTestBlock(httpResponseTestCase);
+        writer.write("\n");
+        // only unmarshall the response and assert if a body is present, as per smithy spec
```
we don't expose the header collection as a property in the response, and unlike other SDKs that may return unmodeled headers in `ResponseMetadata` we do not do that. Unless it is modeled, we don't return the header as part of the unmarshalled response. So unmarshalling the response wouldn't actually do anything in this case as it pertains to headers.

gosar replied:
I'm in fact talking about the case where the header is modeled, but there is no body.

peterrsongg replied:
ah that is true, I changed the condition to only skip unmarshalling if both headers and body is empty

gosar commented on generator/ProtocolTestsGenerator/smithy-dotnet-codegen/src/main/java/software/amazon/smithy/dotnet/codegen/HttpProtocolTestGenerator.java:
```diff
@@ -0,0 +1,551 @@
+            writer.write("var actualResponse = ($L)unmarshalledResponse;", responseSymbol);
+            writer.write("Comparer.CompareObjects<$L>(expectedResponse,actualResponse);", responseSymbol);
+        }
+        if (!httpResponseTestCase.getForbidHeaders().isEmpty()) {
+            for (var forbidHeader : httpResponseTestCase.getForbidHeaders()) {
+                writer.write("Assert.IsFalse(context.ResponseData.IsHeaderPresent($S);", forbidHeader);
+            }
+        }
+        if (!httpResponseTestCase.getRequireHeaders().isEmpty()) {
+            for (var requireHeader : httpResponseTestCase.getRequireHeaders()) {
+                writer.write("Assert.IsTrue(context.ResponseData.IsHeaderPresent($S);", requireHeader);
+            }
+        }
```
sorry, looking at this code I realized I did not think through this fully. I saw the httpResponseTests can have required/forbid headers and didn't see support for it in the generator, so pointed it out. But now I realize that you are only generating client tests. required/forbid headers for httpResponseTests would be something to assert on in server tests. They are not relevant to client tests. So please remove this. Sorry for that.

gosar commented on generator/ProtocolTestsGenerator/smithy-dotnet-protocol-test/smithy-build.json:
```diff
@@ -0,0 +1,131 @@
+{
+  "version": "1.0",
+  "projections": {
```
there are also other service specific tests, e.g., https://github.com/smithy-lang/smithy/tree/main/smithy-aws-protocol-tests/model/restXml/services, https://github.com/smithy-lang/smithy/tree/main/smithy-aws-protocol-tests/model/restJson1/services that should be included.
